### PR TITLE
Run performance tests

### PR DESCRIPTION
**Pull request by lcarva**
**This is very much a draft!**

The main change here is in `models.py` and `api_v1.py` to support bulk operations instead of processing one dependency at a time.

The results for running this against my local development instance is as follows:
| Scenario                 | get_or_create_dep | add_dep           | deps_commit       | deps_overall     |
| --- | --- | --- | --- | --- |
| small_bulk_update_30     | 0.0132            | 0.0209            | 0.0018            | 0.0427            |
| small_bulk_update_50     | 0.0107            | 0.0397            | 0.0019            | 0.0555            |
| small_bulk_update_300    | 0.0104            | 0.0155            | 0.0021            | 0.0313            |
| small_bulk_update_500    | 0.0103            | 0.0153            | 0.0019            | 0.0309            |
| small_bulk_update_1000   | 0.0105            | 0.0150            | 0.0018            | 0.0307            |
| small_single_update_30   | 0.0558            | 0.0730            | 0.0049            | 0.1374            |
| small_single_update_50   | 0.0448            | 0.0602            | 0.0051            | 0.1135            |
| small_single_update_300  | 0.0541            | 0.0760            | 0.0053            | 0.1424            |
| small_single_update_500  | 0.0504            | 0.0663            | 0.0047            | 0.1249            |
| small_single_update_1000 | 0.0605            | 0.0772            | 0.0059            | 0.1478            |
| large_bulk_update_30     | 8.3260            | 11.8212           | 0.7970            | 21.4277           |
| large_bulk_update_50     | 7.2226            | 9.2437            | 0.5015            | 17.3605           |
| large_bulk_update_300    | 6.3200            | 5.8793            | 0.1453            | 12.6846           |
| large_bulk_update_500    | 6.5995            | 5.5716            | 0.1271            | 12.6810           |
| large_bulk_update_1000   | 6.5346            | 5.2714            | 0.0970            | 12.2814           |
| large_single_update_30   | 52.2667           | 67.7917           | 2.0673            | 122.9351          |
| large_single_update_50   | 44.7120           | 57.6657           | 1.2301            | 104.3097          |
| large_single_update_300  | 40.9292           | 52.4725           | 0.3211            | 94.3223           |
| large_single_update_500  | 48.7412           | 61.8422           | 0.2998            | 111.6252          |
| large_single_update_1000 | 41.2431           | 52.6765           | 0.2113            | 94.7579           |

I'd like to run these tests against our stage instance as it will provide more realistic results.

### CODE REVIEW

File: cachito/web/manage.py [hunk 1]
```diff
@@ -31,5 +31,180 @@ def wait_for_db():
             break
 
 
+@cli.command(name="test-performance")
+def test_performance():
```
Comment (lcarva) [inline cachito/web/manage.py:35]: This doesn't really belong here, but it was convenient for testing.

**Comment by arewm** [comment]
How should I interpret the column headers?  If everything is explained in the code, tell me to read it. I just looked at your first comment so far :)

**Comment by lcarva** [comment]
> How should I interpret the column headers? If everything is explained in the code, tell me to read it. I just looked at your first comment so far :)

You can infer it by reading the code. I'll explain it here to make it more obvious :)


- **Scenario**: the name of the test scenario. `small` prefix means a request with just a few dependencies, i.e. [retrodep](https://github.com/release-engineering/retrodep). `large` prefix means a request with many dependencies, i.e. [tekton-pipelines](https://github.com/openshift/tektoncd-pipeline/tree/2ba62c8cfc98b74a6d02e692ce181add8b2184cd). `bulk` vs `single` means bulk operations vs processing one entity at a time. The number at the end is the size of the update batch.
- **get_or_create_dep**: cumulative duration, in seconds, of adding or reusing dependencies in the database
- **add_dep**: cumulative duration, in seconds, of associating dependencies with a request
- **deps_commit**: cumulative duration, in seconds, of committing database transactions after the two operations above
- **deps_overall**: cumulative duration, in seconds, of performing all three operation above

**Comment by athos-ribeiro** [comment]
Nice results!

Thanks for the explanation on the scenarios as well.

+1 to testing it in stage. +1 to push this to a new branch (quay should build a new image tagged after the new git branch) for stage testing.

File: cachito/web/models.py [hunk 2]
```diff
@@ -189,6 +189,30 @@ def get_or_create(cls, package):
 
         return package_object
 
+    @classmethod
+    def bulk_get_or_create(cls, packages):
+        """
+        TODO
+        """
+        # TODO: Make this dynamic and take alternative, slow path, if not postgres
```
Comment (lcarva) [inline cachito/web/models.py:197]: For future reference:
```python
>>> from sqlalchemy.engine import reflection
>>> from cachito.web import db
>>> insp = reflection.Inspector.from_engine(db.engine)
>>> insp.dialect.name
'postgresql'
```

**Comment by lcarva** [comment]
Dropping this. The final decision has been to not track dependencies in the database at all. Instead, generate a json file associated with a given Cachito request which gets removed once the Cachito request goes stale.

**Pull request closed without merging**